cni/cmd/istio-cni-taint/main.go

```diff
@@ -0,0 +1,189 @@
+	}
+
+	if viper.GetBool("help") {
+		pflag.Usage()
+		os.Exit(0)
+	}
+
+	viper.SetEnvPrefix("TAINT")
+	viper.AutomaticEnv()
+	// Pull runtime args into structs
+	options = &ControllerOptions{
+		RunAsDaemon: viper.GetBool("run-as-daemon"),
+		TaintOptions: &taint.Options{
+			ConfigmapName:      viper.GetString("configmap-name"),
+			ConfigmapNamespace: viper.GetString("configmap-namespace"),
+		},
+	}
+
+	return
+}
+
+// Set up Kubernetes client using kubeconfig (or in-cluster config if no file provided)
+func clientSetup() (clientset *client.Clientset, err error) {
+	loadingRules := clientcmd.NewDefaultClientConfigLoadingRules()
+	configOverrides := &clientcmd.ConfigOverrides{}
+	kubeConfig := clientcmd.NewNonInteractiveDeferredLoadingClientConfig(loadingRules, configOverrides)
+	config, err := kubeConfig.ClientConfig()
+	if err != nil {
+		return
+	}
+	clientset, err = client.NewForConfig(config)
+	return
+}
+
+// Log human-readable output describing the current filter and option selection
+func logCurrentOptions(ts *taint.TaintSetter, options *ControllerOptions) {
+	if options.RunAsDaemon {
+		log.Infof("Controller Option: Running as a Daemon.")
+	}
+	for _, cs := range ts.GetAllConfigs() {
+		log.Infof("ConfigSetting %s", cs.ToString())
+	}
+}
+
+//check all node, taint all unready node
+func nodeReadinessCheck(tc *taint.Controller) {
+	nodes := tc.ListAllNode()
+	for _, node := range nodes {
+		err := tc.ProcessNode(node)
+		if err != nil {
+			log.Fatalf("error: %+v in node %v", err.Error(), node.Name)
+		}
+	}
+}
+
+func main() {
+	loggingOptions.OutputPaths = []string{"stderr"}
+	loggingOptions.JSONEncoding = true
+	if err := log.Configure(loggingOptions); err != nil {
+		os.Exit(1)
+	}
+	options := parseFlags()
+
+	clientSet, err := clientSetup()
+	if err != nil {
+		log.Fatalf("Could not construct clientSet: %s", err)
+	}
+	taintSetter, err := taint.NewTaintSetter(clientSet, options.TaintOptions)
+	if err != nil {
+		log.Fatalf("Could not construct taint setter: %s", err)
+	}
+	logCurrentOptions(&taintSetter, options)
+	tc, err := taint.NewTaintSetterController(&taintSetter)
+	if err != nil {
+		log.Fatalf("Fatal error constructing taint controller: %+v", err)
+	}
+	if options.RunAsDaemon {
+		id := uuid.New().String()
+		stopCh := make(chan struct{})
+		lock := &resourcelock.LeaseLock{
+			LeaseMeta: metav1.ObjectMeta{
+				Name:      LeassLockName,
+				Namespace: LeaseLockNamespace,
+			},
+			Client: clientSet.CoordinationV1(),
+			LockConfig: resourcelock.ResourceLockConfig{
+				Identity: id,
+			},
+		}
+		leadelectionCallback := leaderelection.LeaderCallbacks{
+			OnStartedLeading: func(ctx context.Context) {
+				//once leader elected it should taint all nodes at first to prevent race condition
+				tc.RegistTaints()
+				tc.Run(ctx.Done()) //graceful shut down
+			},
+			OnStoppedLeading: func() {
+				// when leader failed, log leader failure and restart leader election
+				log.Infof("leader lost: %s", id)
```

Comment:
Have you tested that this actually restarts leader election? we have a lot of bugs around this..

Reply:
I have tested  it by running multiple instances and kill leader, and I found that they can cancel taint controller at first and then restart leader election